File: src/main/java/javaversion/UserController.java
package javaversion;

import java.util.*;

public class UserController {
    private final Database database;
    private final Mailer mailer;

    public UserController(Database database, Mailer mailer) {
        this.database = Objects.requireNonNull(database);
        this.mailer = Objects.requireNonNull(mailer);
    }

    public void changeEmail(String userId, String newEmail) {
        var user = database.getUserById(userId)
            .orElseThrow(() -> new RuntimeException("ユーザーが見つかりませんでした。ユーザーID: " + userId));

        var email = user.get("email");

        if (email.equals(newEmail)) {
            return;
        }

        var company = database.getCompany();
        var companyDomainName = company.get("companyDomainName");
        var numberOfEmployees = Integer.parseInt(company.get("numberOfEmployees"));

        var emailDomain = newEmail.split("@")[1];
        var newType = emailDomain.equals(companyDomainName) ? "EMPLOYEE" : "CUSTOMER";

        var delta = newType.equals("EMPLOYEE") ? 1 : -1;
        var newNumber = numberOfEmployees + delta;

        var newCompany = new HashMap<>(company);
        newCompany.put("numberOfEmployees", String.valueOf(newNumber));
        newCompany.put("companyDomainName", emailDomain);
        database.saveCompany(newCompany);

        var newUser = new HashMap<>(user);
        newUser.put("email", newEmail);
        newUser.put("userType", newType);

        database.saveUser(newUser);
        mailer.sendEmailChangedMessage(userId, newEmail);
    }
}

File: src/main/kotlin/UserController.kt
class UserController(
    private val database: Database,
    private val mailer: Mailer,
) {

    fun changeEmail(userId: String, newEmail: String) {
        val user = database.getUserById(userId)!!

        val email = user["email"]!!
        val userType = user["userType"]!!

        if (email == newEmail) {
            return
        }

        val company = database.getCompany()
        val companyDomainName = company["companyDomainName"]!!
        val numberOfEmployees = company["numberOfEmployees"]!!.toInt()

        val emailDomain = newEmail.split("@")[1]

        val newType = if (emailDomain == companyDomainName) "EMPLOYEE" else "CUSTOMER"

        if (userType != newType) {
            val delta = if (newType == "EMPLOYEE") 1 else -1
            val newNumber = numberOfEmployees + delta
            company["numberOfEmployees"] = newNumber.toString()
            company["companyDomainName"] = emailDomain
            database.saveCompany(company)
        }

        user["email"] = newEmail
        user["userType"] = newType

        database.saveUser(user)
        mailer.sendEmailChangedMessage(userId, newEmail)
    }
}
